ksql-api/src/main/java/io/confluent/ksql/api/server/InsertsBodyParser.java

```diff
@@ -0,0 +1,93 @@
+import static io.confluent.ksql.api.server.ServerUtils.handleError;
+
+import io.confluent.ksql.api.spi.Endpoints;
+import io.confluent.ksql.api.spi.InsertsSubscriber;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * This class handles the parsing of the request body for a stream of inserts. The user can send a
+ * stream of inserts to the server by opening an HTTP2 request to the server at the appropriate uri
+ * and POSTing a request. The request must contain an initial JSON object (encoded as UTF-8 text)
+ * which contains the arguments for the request (e.g. the target to insert into, and whether acks
+ * are wanted) This must be followed by a new line, and then followed by zero or more JSON objects
+ * (also encoded as UTF-8 text) each representing a row to insert. The last JSON object must be
+ * followed by a new-line.
+ */
+public class InsertsBodyParser {
+
+  private final Endpoints endpoints;
+  private final RoutingContext routingContext;
+  private boolean readArguments;
+  private InsertsPublisher publisher;
+  private long rowsReceived;
+  private AcksSubscriber acksSubscriber;
+
+  public InsertsBodyParser(final Endpoints endpoints, final RoutingContext routingContext) {
+    this.endpoints = endpoints;
+    this.routingContext = routingContext;
+    routingContext.response().endHandler(v -> {
+      if (publisher != null) {
+        publisher.close();
+      }
+    });
+  }
+
+  public void handleBodyEnd(final Void v) {
+    if (acksSubscriber == null) {
+      routingContext.response().end();
+    } else {
+      // We close the response after the stream of acks has been sent
+      acksSubscriber.insertsSent(rowsReceived);
+    }
+  }
+
+  public void handleBodyBuffer(final Buffer buff) {
+    if (!readArguments) {
+      JsonObject args = new JsonObject(buff);
+      readArguments = true;
+      String target = args.getString("target");
+      if (target == null) {
+        handleError(routingContext.response(), 400, ERROR_CODE_MISSING_PARAM,
+            "No target in arguments");
+        return;
+      }
+      Boolean acks = args.getBoolean("acks");
+      if (acks == null) {
+        handleError(routingContext.response(), 400, ERROR_CODE_MISSING_PARAM,
+            "No acks in arguments");
+        return;
+      }
+      JsonObject properties = args.getJsonObject("properties");
+      routingContext.request().endHandler(this::handleBodyEnd);
+      acksSubscriber = acks ? new AcksSubscriber(routingContext.response()) : null;
+      InsertsSubscriber insertsSubscriber = endpoints
+          .createInsertsSubscriber(target, properties, acksSubscriber);
+      publisher = new InsertsPublisher();
+      publisher.subscribe(insertsSubscriber);
+    } else if (publisher != null) {
+      JsonObject row = new JsonObject(buff);
+      publisher.receiveRow(row);
+      rowsReceived++;
+    }
+  }
```

Comment:
As you probably know, I tend to err on the side of minimalism and avoid proliferation of classes, especially for simple things. In this case we're just extracting a couple of fields from a map so I did the simplest possible thing.
I think it would be possible to use a Jackson ObjectMapper and create POJOs for the arguments (although it's not obvious to me exactly how I would do this, especially considering the params to query-stream contains a nested JsonObject in the properties field), but it would certainly result in a lot more code. Considering these args classes would never be used outside this specific method, my initial thoughts are this would be overkill. If the classes were passed to other parts of the program, I think the case would be more clear cut.

Reply:
I will however look into how hard this would be to do. One other thing to bear in mind that Vert. uses JsonObject in many APIs - it's kind of the "Vert.x-way" to do JSON. One of the reasons for that is also performance.

Reply:
https://github.com/confluentinc/ksql/issues/4333

Reply:
Adding my thoughts here - I tend to agree with Andy on this one. If there's a legitimate concern on performance than perhaps JSON objects that are used exclusively internally and are not exposed can be hand crafted - but it would be nice to have them constructed in a centralized place so that we can define the "API" coherently.

On the other hand, anything that is exposed externally I feel should definitely have an associated POJO to define the public API and allow us to automatically generate documentation on it.

Reply:
If you're looking for examples of how to use Pojos with Json then take a look at the pojos in the current API are used; they're in the `io.confluent.ksql.rest.entity` package.

the general pattern is that you need a Jackson `ObjectMapper` and it can serialize a pojo to JSON and deserialize the JSON to a pojo.

Ideally, put the model into a different module, so that it can be shared between client and server.

Actually, that makes me think: we're looking at clients with different languages, so what we really need for the API is the JSON schema.  We can then use this to generate / validate clients.   We can generate the schema from the Java pojos: @Rohan has done this for our internal command API.  Alternatively, you may wish to hand write the JSON schema, from which we may be able to generate the Pojos.

Thoughts?

Reply:
Chatted with Andy about this privately. Here's what code would like if we had to deserialize into POJOs and catch various exceptions so we can send back the right error codes on the wire: https://gist.github.com/purplefox/8975123fcc97da9c7f00eb2ce5ccf8ec
Super ugly I hope you agree. Now, sure we can hide this in a utility class but still.